crates/oxc_linter/src/rules/eslint/no_nonoctal_decimal_escape.rs

```diff
@@ -0,0 +1,245 @@
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::{self, Error},
+};
+use oxc_macros::declare_oxc_lint;
+use oxc_span::Span;
+use regex::{Captures, Match, Regex};
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+#[derive(Debug, Error, Diagnostic)]
+enum NoNonoctalDecimalEscapeDiagnostic {
+    #[error("eslint(no-nonoctal-decimal-escape): Don't use '{0}' escape sequence.")]
+    #[diagnostic(
+        severity(warning),
+        help("Replace '{0}' with '{1}'. This maintains the current functionality.")
+    )]
+    Replacement(String, String, #[label] Span),
+
+    #[error("eslint(no-nonoctal-decimal-escape): Don't use '{0}' escape sequence.")]
+    #[diagnostic(
+        severity(warning),
+        help("Replace '{0}' with '{1}' to include the actual backslash character.")
+    )]
+    EscapeBackslash(String, String, #[label] Span),
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct NoNonoctalDecimalEscape;
+
+declare_oxc_lint!(
+    /// ### What it does
+    /// This rule disallows \8 and \9 escape sequences in string literals
+    ///
+    /// ### Why is this bad?
+    /// ECMAScript specification treats \8 and \9 in string literals as a legacy feature
+    ///
+    /// ### Example
+    /// ```javascript
+    /// incorrect:
+    /// "\8"
+    /// "\9"
+    /// correct:
+    /// "8"
+    /// "\\9"
+    /// ```
+    NoNonoctalDecimalEscape,
+    correctness
+);
+
+impl Rule for NoNonoctalDecimalEscape {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        if let AstKind::StringLiteral(literal) = node.kind() {
+            check_string(ctx, literal.span.source_text(ctx.source_text()));
+        }
+    }
+}
+trait StickyRegex {
+    // takes the start position and returns the match and the nextPosition
+    fn sticky_find<'h>(&self, haystack: &'h str, start: usize) -> (Option<Match<'h>>, usize);
+    fn sticky_captures<'h>(&self, haystack: &'h str, start: usize)
+        -> (Option<Captures<'h>>, usize);
+}
+
+impl StickyRegex for Regex {
+    fn sticky_find<'h>(&self, haystack: &'h str, start: usize) -> (Option<Match<'h>>, usize) {
+        let m_opt = self.find_at(haystack, start);
+        if let Some(m) = m_opt {
+            if m.start() == start {
+                return (m_opt, m.end());
+            }
+        }
+        (None, 0)
+    }
+
+    fn sticky_captures<'h>(
+        &self,
+        haystack: &'h str,
+        start: usize,
+    ) -> (Option<Captures<'h>>, usize) {
+        let c_opt = self.captures_at(haystack, start);
+        if let Some(captures) = c_opt {
+            let capture = captures.get(0).unwrap();
+            if capture.start() == start {
+                return (Some(captures), capture.end());
+            }
+        }
+        (None, 0)
+    }
+}
+
+#[allow(clippy::cast_possible_truncation)]
+fn check_string(ctx: &LintContext<'_>, string: &str) {
+    lazy_static! {
+        static ref QUICK_TEST_REGEX: Regex = Regex::new(r"\\[89]").unwrap();
```

Comment:
I think we can do a quick scan with string methods instead of using a regex, which is quite heavy.